dags/iop/xml_extractors.py

```diff
@@ -0,0 +1,112 @@
+import xml.etree.ElementTree as ET
+
+from common.parsing.extractor import IExtractor
+from structlog import get_logger
+
+
+class TextExtractor(IExtractor):
+    def __init__(
+        self,
+        destination,
+        source,
+        required=True,
+        default_value=None,
+        extra_function=lambda s: s,
+        prefixes=None,
+    ) -> None:
+        super().__init__(destination)
+        self.destination = destination
+        self.source = source
+        self.prefixes = prefixes
+        self.required = required
+        self.default_value = default_value
+        self.extra_function = extra_function
+        self.logger = get_logger().bind(class_name=type(self).__name__)
+
+    def extract(self, article: ET.Element):
+        if self.prefixes:
+            node_with_prefix = self.extra_function(
+                article.find(self.source, self.prefixes).text
+            )
+            return node_with_prefix
+        node = article.find(self.source)
+        # if the value is required, it cannot be empty array or [None]
+        if bool(node) and all(node) and self.required:
+            return self.extra_function(article.find(self.source).text)
+        elif not self.required and node:
+            return self.extra_function(article.find(self.source).text)
+        elif self.required and node is None:
+            raise RequiredFieldNotFoundExtractionError(self.destination)
+        elif not bool(self.default_value) and self.required:
+            raise RequiredFieldNotFoundExtractionError(self.destination)
+        return self.default_value
+
```

Comment:
it's better to use the guard clauses technique for these extractors https://www.davidhayden.me/blog/guard-clauses-in-computer-science

Reply:
I am not sure if I am managed to use it correctly